### PR TITLE
Adds support to allow usage of none-cloneable MAC providers

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description

`MacSpi` implements Cloneable interface but according to documentation not all `MacSpi` providers need to support cloning. It is possible that attempt to clone throws `CloneNotSupportedException`. The MacPool implementation in the Cosmos SDK needs to gracefully handle this instead of mapping to `IllegalStateException`. This is a customer reported incident and will need a hotfix as soon as possible.

Adding a section with some context for the "id" - eoncoding unit tests also added in this PR

Got reports that certain valid characters in id were not handled appropriately in node.js SDK (space at the end of id for example). When testing I realized inconsistency across SDKs and even Gateway vs. Direct - so I will go through and create a baseline with end-to-end tests in all SDKs and then see how we can fix this situation (and whether we think it is important to fix vs. just better and more clear document the various constraints)

Overview is here: [https://microsofteur-my.sharepoint.com/❌/g/personal/fabianm_microsoft_com/EdA2zPvwjfxCo05SKu4FQXMBpnt_8HpDgSIOxWZqpjkikQ?e=6hX4ud](https://microsofteur-my.sharepoint.com/:x:/g/personal/fabianm_microsoft_com/EdA2zPvwjfxCo05SKu4FQXMBpnt_8HpDgSIOxWZqpjkikQ?e=6hX4ud)

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemIdEncodingTest.java [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+import org.testng.annotations.Ignore;
+import org.testng.annotations.Test;
+import reactor.core.Exceptions;
+
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+public class CosmosItemIdEncodingTest extends TestSuiteBase {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemIdEncodingTest.java:31]: should this belong to a different PR?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java [hunk 2]
```diff
@@ -3,32 +3,64 @@
...
+        // But while Mac implements Cloneable interface not every MacSpi is required to support cloning
+        // If it doesn't  - bad for performance - but functionally correct and necessary
+        // - see https://docs.oracle.com/javase/7/docs/technotes/guides/security/crypto/HowToImplAProvider.html#Step11
+        // - see https://docs.oracle.com/en/java/javase/18/docs/api/java.base/javax/crypto/Mac.html#clone()
+        try {
+            return (Mac) this.macInstance.clone();
         } catch (CloneNotSupportedException e) {
-            throw new IllegalStateException(e);
+
+            LOGGER.warn(
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java:54]: Do we need this warning log?
I am concerned that customers might see this log and blame for perf issues, while at the same time not sure on how much control they have got on this?

**Review by kushagraThapar — APPROVED** [review]
